File: Easy/PYTHON/RemoveDuplicatesFromLinkedList.py
from typing import Counter


class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
    
    def add_many(self, values):
        current = self
        while current.next is not None:
            current = current.next
        
        for value in values:
            current.next = Node(value)
            current = current.next
        
        return self

    def get_nodes(self):
        nodes = []
        current = self
        while current is not None:
            nodes.append(current.value)
            current = current.next
        
        return nodes

def remove_duplicates_from_linkedlist(node):
    current = node
    while current is not None and  current.next is not None:
        nextUniqueNode =  current.next
        while nextUniqueNode is not None and nextUniqueNode.value == current.value:
            nextUniqueNode = nextUniqueNode.next
            
        current.next = nextUniqueNode
        current = nextUniqueNode
                                           
    return node


input = Node(1)
values = [1, 1, 2, 3, 8, 8, 8, 9, 12]
input.add_many(values)

before = input.get_nodes()
print(before)

remove_duplicates_from_linkedlist(input)

after = input.get_nodes()
print(after)

File: Easy/PHP/RemoveDuplicatesFromLinkedList.php
<?php

namespace Easy\PHP\RemoveRemoveDuplicatesFromLinkedList;

class Node
{
    public $value;
    public $next;

    public function __construct($value)
    {
        $this->value = $value;
        $this->next = null;
    }

    public function getNodes()
    {
        $nodes = [];
        $current = $this;
        while ($current != null) {
            array_push($nodes, $current->value);
            $current = $current->next;
        }

        return $nodes;
    }

    public function addMany($values)
    {
        $current = $this;
        while ($current->next != null) {
            $current = $current->next;
        }

        foreach ($values as $value) {
            $current->next = new Node($value);
            $current = $current->next;
        }

        return $this;
    }
}

function removeDuplicatesFromLinkedList($node)
{
    $current = $node;
    while ($current != null && $current->next != null) {
        $nextUniqueNode = $current->next;
        while ($nextUniqueNode != null &&  $nextUniqueNode->value == $current->value) {
            $nextUniqueNode =  $nextUniqueNode->next;
        }
        $current->next =  $nextUniqueNode;
        $current =  $nextUniqueNode;
    }

    return $node;
}

$input = new Node(1);
$values = [1, 1, 2, 3, 8, 8, 8, 9, 12];
$input->addMany($values);

$before = $input->getNodes();
print_r($before);

removeDuplicatesFromLinkedList($input, $values);

$after = $input->getNodes();
print_r($after);
